packages/jest-resolve/src/shouldLoadAsEsm.ts

```diff
@@ -68,11 +69,21 @@ function shouldLoadAsEsm(path: Config.Path): boolean {
 
 function cachedPkgCheck(cwd: Config.Path): boolean {
   // TODO: can we cache lookups somehow?
-  const pkg = readPkgUp.sync({cwd, normalize: false});
-
-  if (!pkg) {
+  const pkgPath = escalade(cwd, (_dir, names) => {
+    if (names.includes('package.json')) {
+      // will be resolved into absolute
+      return 'package.json';
+    }
+    return false;
+  });
+  if (!pkgPath) {
     return false;
   }
 
-  return pkg.packageJson.type === 'module';
+  try {
+    const pkg = JSON.parse(readFileSync(pkgPath, 'utf-8'));
```

Comment:
could you create a `Map` of `/path/to/package.json, boolean` so we don't have to read from disk and parse it if we've already looked at the file? Also make sure to clear the map in the function for that in this file